File: python_tests/utils/suite.py
import os
import subprocess
import sys


class Suite(object):
    def __init__(self, directory: str, exclude_file: str):
        self.directory = directory
        self.exclude_file = exclude_file

    def run(self):
        # Iterate the directory, finding files viable for execution
        for root, directories, _ in os.walk(self.directory):
            for directory in directories:
                main_file_path = os.path.join(root, directory, "main.py")
                if not os.path.isfile(main_file_path):
                    continue

                # Run the file in a new subprocess to reduce likelihood of code cache misses
                try:
                    env = os.environ.copy()
                    env["PYTHONPATH"] = os.getcwd()

                    subprocess.run(
                        [sys.executable, main_file_path],
                        cwd=root,
                        env=env,
                        text=True,
                        check=True,
                        stdout=sys.stdout,
                        stderr=sys.stderr,
                    )
                except subprocess.CalledProcessError as e:
                    print(f"Error running {main_file_path}:\n{e.stderr}", file=sys.stderr)

File: go_tests/utils/suite.go
package utils

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

type Suite struct {
	directory string
}

func NewSuite(directory string) (instance *Suite) {
	return &Suite{
		directory: directory,
	}
}

func (instance *Suite) Run() (err error) {
	directories, err := os.ReadDir(instance.directory)
	if err != nil {
		return err
	}

	// Iterate the directory, finding files viable for execution
	for _, directory := range directories {
		if !directory.IsDir() {
			continue
		}

		// Check if the directory contains a main.go file
		mainFilePath := filepath.Join(instance.directory, directory.Name(), "main.go")
		_, err = os.Stat(mainFilePath)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}

			return fmt.Errorf("failed to stat %s: %w", mainFilePath, err)
		}

		// Compile the main.go file
		executablePath := directory.Name() + ".out"
		compileCmd := exec.Command("go", "build", "-o", executablePath, mainFilePath)
		compileCmd.Stdout = os.Stdout
		compileCmd.Stderr = os.Stderr

		err = compileCmd.Run()
		if err != nil {
			return fmt.Errorf("failed to compile %s: %w", mainFilePath, err)
		}
		defer os.Remove(executablePath)

		// Run the compiled executable
		runCmd := exec.Command("./" + executablePath)
		runCmd.Stdout = os.Stdout
		runCmd.Stderr = os.Stderr

		err = runCmd.Run()
		if err != nil {
			return fmt.Errorf("failed to run %s: %w", executablePath, err)
		}
	}

	return nil
}
